File: asr/harmony/sdk/src/main/ets/com/amphion/asr/StreamingAgcProcessor.ts
export interface AgcBackend {
  process(frame: Float32Array): Float32Array;
  close(): void;
}

export type AgcBackendFactory = () => AgcBackend;

export class ProcessedAudioFrame {
  raw: Float32Array;
  processed: Float32Array;

  constructor(raw: Float32Array, processed: Float32Array) {
    this.raw = raw;
    this.processed = processed;
  }
}

/** Reframes arbitrary caller chunks into the 10 ms mono frames required by WebRTC AGC2. */
export class StreamingAgcProcessor {
  private frameSamples: number;
  private backendFactory: AgcBackendFactory;
  private backend?: AgcBackend;
  private carry: Float32Array = new Float32Array(0);
  private closed: boolean = false;

  constructor(sampleRate: number, backendFactory: AgcBackendFactory) {
    if (sampleRate <= 0 || sampleRate % 100 !== 0) {
      throw new Error(`sampleRate must have an integral 10ms frame, got ${sampleRate}`);
    }
    this.frameSamples = sampleRate / 100;
    this.backendFactory = backendFactory;
  }

  process(samples: Float32Array): ProcessedAudioFrame[] {
    if (this.closed) throw new Error('AGC processor is closed');
    if (samples.length === 0) return [];
    const merged = concat(this.carry, samples);
    const completeSamples = Math.floor(merged.length / this.frameSamples) * this.frameSamples;
    if (completeSamples === 0) {
      this.carry = merged.slice();
      return [];
    }
    const rawOutput = new Float32Array(completeSamples);
    const processedOutput = new Float32Array(completeSamples);
    let offset = 0;
    while (offset < completeSamples) {
      const raw = merged.slice(offset, offset + this.frameSamples);
      const frame = this.processFrame(raw);
      rawOutput.set(frame.raw, offset);
      processedOutput.set(frame.processed, offset);
      offset += this.frameSamples;
    }
    this.carry = offset < merged.length ? merged.slice(offset) : new Float32Array(0);
    // AGC still runs on strict 10 ms frames, but preserve the caller's decoder submission
    // granularity. Decoding every internal frame separately halves burst throughput because the
    // Harmony adapter awaits each native decode on its serial audio dispatcher.
    return [new ProcessedAudioFrame(rawOutput, processedOutput)];
  }

  flush(): ProcessedAudioFrame[] {
    if (this.closed) throw new Error('AGC processor is closed');
    if (this.carry.length === 0) return [];

    const raw = this.carry;
    this.carry = new Float32Array(0);
    const padded = new Float32Array(this.frameSamples);
    padded.set(raw);
    const processed = this.processFrame(padded).processed.slice(0, raw.length);
    return [new ProcessedAudioFrame(raw, processed)];
  }

  close(): void {
    if (this.closed) return;
    this.closed = true;
    this.carry = new Float32Array(0);
    this.backend?.close();
    this.backend = undefined;
  }

  private processFrame(raw: Float32Array): ProcessedAudioFrame {
    if (this.backend === undefined) this.backend = this.backendFactory();
    const processed = this.backend.process(raw);
    if (processed.length !== raw.length) {
      throw new Error(`AGC backend changed frame size: ${raw.length} -> ${processed.length}`);
    }
    return new ProcessedAudioFrame(raw, processed);
  }
}

function concat(left: Float32Array, right: Float32Array): Float32Array {
  if (left.length === 0) return right;
  const output = new Float32Array(left.length + right.length);
  output.set(left, 0);
  output.set(right, left.length);
  return output;
}

File: asr/android/sdk/src/main/java/com/amphion/asr/internal/StreamingAgcProcessor.kt
package com.amphion.asr.internal

import java.io.Closeable

internal data class ProcessedAudioFrame(
    val raw: FloatArray,
    val processed: FloatArray,
)

internal interface AgcBackend : Closeable {
    /** Process exactly one 10 ms mono PCM frame and return the same sample count. */
    fun process(frame: FloatArray): FloatArray
}

/**
 * Converts arbitrary caller chunks into the fixed 10 ms frames required by WebRTC AGC2.
 *
 * Raw and processed samples stay paired so VAD and speaker verification keep using the original
 * waveform while only ASR consumes the normalized waveform. This object is decoder-thread only.
 */
internal class StreamingAgcProcessor(
    sampleRate: Int,
    private val backendFactory: () -> AgcBackend = { NativeAgcBackend(sampleRate) },
) : AgcFrameProcessor {

    private val frameSamples = (sampleRate / FRAMES_PER_SECOND).also {
        require(sampleRate > 0 && sampleRate % FRAMES_PER_SECOND == 0) {
            "sampleRate must have an integral 10ms frame, got $sampleRate"
        }
    }
    private var carry = FloatArray(0)
    private var backend: AgcBackend? = null
    private var closed = false

    override fun process(samples: FloatArray): List<ProcessedAudioFrame> {
        check(!closed) { "AGC processor is closed" }
        if (samples.isEmpty()) return emptyList()
        val merged = if (carry.isEmpty()) samples else carry + samples
        val completeSamples = merged.size / frameSamples * frameSamples
        if (completeSamples == 0) {
            carry = merged.copyOf()
            return emptyList()
        }
        val rawOutput = FloatArray(completeSamples)
        val processedOutput = FloatArray(completeSamples)
        var offset = 0
        while (offset < completeSamples) {
            val raw = merged.copyOfRange(offset, offset + frameSamples)
            val frame = processFrame(raw)
            frame.raw.copyInto(rawOutput, offset)
            frame.processed.copyInto(processedOutput, offset)
            offset += frameSamples
        }
        carry = if (offset < merged.size) merged.copyOfRange(offset, merged.size) else FloatArray(0)
        // Keep WebRTC's 10 ms processing contract internal. The decoder should retain the caller's
        // submission granularity instead of paying one decode dispatch per AGC frame.
        return listOf(ProcessedAudioFrame(rawOutput, processedOutput))
    }

    override fun flush(): List<ProcessedAudioFrame> {
        check(!closed) { "AGC processor is closed" }
        if (carry.isEmpty()) return emptyList()

        val raw = carry
        carry = FloatArray(0)
        val padded = raw.copyOf(frameSamples)
        val processed = processFrame(padded).processed.copyOf(raw.size)
        return listOf(ProcessedAudioFrame(raw, processed))
    }

    private fun processFrame(raw: FloatArray): ProcessedAudioFrame {
        val processed = (backend ?: backendFactory().also { backend = it }).process(raw)
        check(processed.size == raw.size) {
            "AGC backend changed frame size: ${raw.size} -> ${processed.size}"
        }
        return ProcessedAudioFrame(raw, processed)
    }

    override fun close() {
        if (closed) return
        closed = true
        carry = FloatArray(0)
        backend?.close()
        backend = null
    }

    private companion object {
        const val FRAMES_PER_SECOND = 100
    }
}
